# Jawablah poin-poin assignment berikut
#### perhatikan dengan baik apa yang diminta pada soal

# 1. Anda bekerja sebagai seorang data scientist, atasan anda mengirimkan anda data kepada anda dan meminta untuk memberikan analisa dengan menggunakan semua metode pada data manipulation berikut :
- a. Filtering
- b. Sorting
- c. Grouping / Aggregasi
- d. Join / Merging
- e. Pivot
- f. Crosstab
- g. Appending
- h. Mengeksplor terkait dengan operasi pada Datetime (week, day, dan sebagainya)
### Notes : Sebelum menjawab permintaan diatas, pertama-tama awali dengan mengajukan pertanyaan analisa terlebih dahulu dan tulis insight dari yang anda dapat, **pertanyaan analisa dilakukan pada semua metode data manipulation diatas**

# Jawaban Soal 1: Pertanyaan Analisa Pada Metode Filtering

- Berapa jumlah transaksi yang menggunakan metode pembayaran E-Wallet, dan apakah pengguna E-Wallet memiliki tingkat kepuasan (Rating) yang lebih tinggi dibandingkan rata-rata keseluruhan? Aku ingin menganalisis, apakah terdapat peningkatan tingkat kepuasan pelanggan (Rating)?

# Jawaban Soal 1: Coding Untuk Metode Filtering

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd

In [4]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


In [5]:
# Langkah pertama kita mau melakukan proses filtering dengan cara mengambil setiap transaksi dari customer yang hanya memakai
# metode pembayaran dengan E-Wallet.
transaksi_ewallet = data_penjualan[data_penjualan['Payment'] == 'Ewallet']

# Langkah kedua kita mau melakukan proses filtering terhadap setiap transaksi dari customer yang tidak memakai metode pembayaran
# dengan E-Wallet.
transaksi_bukan_ewallet = data_penjualan[data_penjualan['Payment'] != 'Ewallet']

# Langkah ketiga kita melakukan pengambilan jumlah data transaksi yang dilakukan oleh customer menggunakan metode E-Wallet.
jumlah_transaksi_ewallet = transaksi_ewallet.shape[0]

# Langkah keempat kita melakukan perhitungan rata-rata terhadap metode pembayaran E-Wallet yang dipakai customer untuk bertransaksi.
mean_rating_ewallet = transaksi_ewallet['Rating'].mean()

# Langkah kelima kita melakukan perhitungan rata-rata terhadap metode pembayaran yang tidak menggunakan E-Wallet yang dipakai customer
# untuk bertransaksi.
mean_rating_bukan_ewallet = transaksi_bukan_ewallet['Rating'].mean()

# Langkah keenam kita melakukan perhitungan rata-rata rating keseluruhan dari dataset.
mean_rating_semua = data_penjualan['Rating'].mean()

# Langkah ketujuh kita mau menampilkan hasil analisisnya.
print("Proses Analisis:")
print(f"Jumlah Transaksi Yang Memakai E-Wallet: {jumlah_transaksi_ewallet:,} Transaksi")
print(f"Rata-Rata Rating Yang Memakai E-Wallet: {mean_rating_ewallet:.2f}")
print(f"Rata-Rata Rating Yang Tidak Memakai E-Wallet: {mean_rating_bukan_ewallet:.2f}")
print(f"Rata-Rata Rating Keseluruhan: {mean_rating_semua:.2f}")

print("\nHasil Analisis:")
if mean_rating_ewallet > mean_rating_bukan_ewallet:
  print(f"Customer E-Wallet Memiliki Rating Yang Lebih Tinggi ({mean_rating_ewallet - mean_rating_bukan_ewallet:.2f} poin lebih tinggi)")
elif mean_rating_ewallet < mean_rating_bukan_ewallet:
  print(f"Customer E-Wallet Memiliki Rating Yang Lebih Rendah ({mean_rating_bukan_ewallet - mean_rating_ewallet:.2f} poin lebih rendah)")
else:
  print("Tidak Perbedaan Signifikan Pada Rating Antara Customer E-Wallet dengan Customer Non E-Wallet")

Proses Analisis:
Jumlah Transaksi Yang Memakai E-Wallet: 345 Transaksi
Rata-Rata Rating Yang Memakai E-Wallet: 6.95
Rata-Rata Rating Yang Tidak Memakai E-Wallet: 6.99
Rata-Rata Rating Keseluruhan: 6.97

Hasil Analisis:
Customer E-Wallet Memiliki Rating Yang Lebih Rendah (0.04 poin lebih rendah)


# Hasil Kesimpulan:
- Proses Analisis Soal 1: Metode Filtering
- Pertanyaan Analisis: Berapa jumlah transaksi yang menggunakan metode pembayaran E-Wallet, dan apakah pengguna E-Wallet memiliki tingkat kepuasan (Rating) yang lebih tinggi dibandingkan rata-rata keseluruhan? Aku ingin menganalisis, apakah terdapat peningkatan tingkat kepuasan pelanggan (Rating)?

1. Setelah melakukan proses analisis dengan metode filtering terhadap pertanyaan analisis yang sudah dibuat sebelumnya, dapat ditarik kesimpulan bahwa customer yang bertransaksi dengan metode E-Wallet memiliki rata-rata rating yang lebih rendah dengan metode selain E-Wallet dan keseluruhan rating terhadap semua jenis metode pembayaran. Dari hal tersebut menunjukkan bahwa bertransaksi dengan metode pembayaran E-Wallet belum bisa memberikan pengalaman yang memuaskan untuk customer, sehingga disarankan bagi pengelola supermarket untuk evaluasi menyeluruh terhadap fitur-fitur yang ada di metode E-Wallet ini seperti fitur pembaca kartu pada mesin pembayaran E-Wallet apakah berjalan dengan baik atau ada masalah teknis pada mesinnya yang harus diperbaiki, diharapkan dari proses evaluasi tersebut bisa meninggkatkan rata-rata rating pelanggan yang berbelanja menggunakan metode E-Wallet di supermarket tersebut.

# Jawaban Soal 2: Pertanyaan Analisa Pada Metode Sorting

- Berapa distribusi total penjualan produk (Beauty and Health) dan (Sports and Travel) untuk customer pria per bulan? Aku ingin menganalisis, adakah pola kecenderungan terhadap customer pria dalam menentukan self reward yang dilakukan tiap bulan tertentu setelah gajian, apakah dia lebih suka kegiatan outdoor dengan berpergian bersama temannya menikmati alam dengan melakukan pembelian terhadap produk (Sports and Travel) atau dia sukanya kegiatan indoor banyak menghabiskan waktu di kosannya mungkin dengan melakukan pembelian terhadap produk (Beauty and Health)?


# Jawaban Soal 2: Coding Untuk Metode Sorting

In [6]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


In [7]:
# Langkah pertama kita mau mengkonversi kolom 'Date' ke dalam format DateTime agar bisa di ekstrak datanya berdasarkan bulan dan tahun.
data_penjualan['Date'] = pd.to_datetime(data_penjualan['Date'])

# Langkah kedua kita mau ekstrak ke dalam periode bulan dan tahun dalam format YYYY-MM supaya nantinya bisa di grouping per bulan datanya.
data_penjualan['Bulan'] = data_penjualan['Date'].dt.to_period('M')

# Langkah ketiga kita mau filter data customer hanya pria dan dua jenis produk yang sedang dianalisis (Health and Beauty) dan (Sports and Travel).
produk_pilihan_pria = ['Health and beauty', 'Sports and travel']
data_filter_pria = data_penjualan[(data_penjualan['Gender'] == 'Male') &
                             (data_penjualan['Product line'].isin(produk_pilihan_pria))]

# Langkah keempat kita mau melakukan grouping untuk menghitung total penjualan per bulan dan juga untuk total penjualan per lini produk.
penjualan_per_bulan = data_filter_pria.groupby(['Bulan', 'Product line'])['Total'].sum().reset_index()

# Langkah kelima kita mau melakukan sorting untuk mengurutkan data berdasarkan bulan secara ascending (dari yang terkecil ke terbesar)
# dan untuk total penjualan secara descending (dari yang terbesar ke terkecil).
sorting_penjualan = penjualan_per_bulan.sort_values(by=['Bulan', 'Total'], ascending=[True, False])

# Langkah keenam kita mau menampilkan hasil dari proses sorting datanya.
print("Tabel Analisis : \n")
print(sorting_penjualan.to_string(index=False))

# Langkah ketujuh kita mau menghitung total keseluruhan untuk referensi perbandingan antara produk (Health and Beauty) dan (Sports and Travel).
total_health_beauty = penjualan_per_bulan[penjualan_per_bulan['Product line'] == 'Health and beauty']['Total'].sum()
total_sports_travel = penjualan_per_bulan[penjualan_per_bulan['Product line'] == 'Sports and travel']['Total'].sum()

# Langkah kedelapan kita mau menampilkan hasil analisisnya.
print("\nHasil Analisis Minat Customer Pria Terhadap Produk (Health and Beauty) VS (Sports and Travel) Dalam Menentukan Kegiatan Self Reward: ")
print(f"Total penjualan untuk produk (Health and Beauty): {total_health_beauty:,.0f}")
print(f"Total penjualan untuk produk (Sports and Travel): {total_sports_travel:,.0f}")

if total_sports_travel > total_health_beauty:
    print("Customer Pria Lebih Banyak Membeli Produk (Sports and Travel")
elif total_health_beauty > total_sports_travel:
    print("Customer Pria Lebih Banyak Membeli Produk (Health and Beauty)")
else:
    print("Pemilihan Produk (Health and Beauty) Dan Produk (Sports and Travel) Seimbang")

Tabel Analisis : 

  Bulan      Product line      Total
2019-01 Health and beauty 11003.1390
2019-01 Sports and travel 10304.6475
2019-02 Health and beauty  6878.2980
2019-02 Sports and travel  5062.3230
2019-03 Health and beauty 12751.3155
2019-03 Sports and travel 11181.1350

Hasil Analisis Minat Customer Pria Terhadap Produk (Health and Beauty) VS (Sports and Travel) Dalam Menentukan Kegiatan Self Reward: 
Total penjualan untuk produk (Health and Beauty): 30,633
Total penjualan untuk produk (Sports and Travel): 26,548
Customer Pria Lebih Banyak Membeli Produk (Health and Beauty)


# Hasil Kesimpulan
- Proses Analisis Soal 2: Metode Sorting Data
- Pertanyaan Analisis: Berapa distribusi total penjualan produk (Beauty and Health) dan (Sports and Travel) untuk customer pria per bulan? Aku ingin menganalisis, adakah pola kecenderungan terhadap customer pria dalam menentukan self reward yang dilakukan tiap bulan tertentu setelah gajian, apakah dia lebih suka kegiatan outdoor dengan berpergian bersama temannya menikmati alam dengan melakukan pembelian terhadap produk (Sports and Travel) atau dia sukanya kegiatan indoor banyak menghabiskan waktu di kosannya mungkin dengan melakukan pembelian terhadap produk (Beauty and Health)?

1. Setelah melakukan proses analisis dengan metode sorting data terhadap pertanyaan analisis yang sudah dibuat sebelumnya, dapat ditarik kesimpulan bahwa customer pria cenderung lebih banyak membeli produk (Beauty and Health). Hal ini menunjukkan preferensi kegiatan self reward yang dilakukan pria tersebut setiap gajian per bulan adalah menghabiskan waktu dengan melakukan kegiatan indoor (di dalam ruangan) seperti contoh melakukan skincare untuk perawatan pada wajah. Saran dan masukan untuk pengelola supermarket bisa melakukan promo di tiap bulan tertentu bagi produk (Beauty and Health) khusus pria karena melihat tren pembelian terhadap jenis produk tersebut sangat menjanjikan dan diharapkan dari promo tersebut dapat meningkatkan pendapatan terhadap perusahaan.

# Jawaban Soal 3: Pertanyaan Analisa Pada Metode Grouping / Aggregasi

- Berapa rata-rata pengeluaran dan rating terhadap produk (Health and Beauty) pada customer (Pria vs Wanita)? Aku ingin menganalisis produk (Health and Beauty) ini sebenarnya lebih diminati oleh customer pria atau wanita? Sudah sesuai dengan ketentuan soal dengan menggunakan kolom kategorikal "Genre" untuk komparasi customer (Pria vs Wanita), kolom "Product Line" untuk pengujian terhadap produk (Health and Beauty), dan penggunaan kolom numerikal "Total" untuk pengujian terhadap data penjualan dan kolom numerikal "Rating" untuk mengukur kepuasan pelanggan.

# Jawaban Soal 3: Coding Untuk Metode Grouping / Aggregasi

In [8]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


In [9]:
# Langkah pertama kita mau melakukan filtering data hanya untuk produk (Health and Beauty).
produk_health_beauty = data_penjualan[data_penjualan['Product line'] == 'Health and beauty']

# Langkah kedua kita mau melakukan grouping terhadap Gender, selanjutnya dilakukan agregasi pada kolom numerikal.
grouping_gender_customer = produk_health_beauty.groupby('Gender').agg({
    'Total'     : ['mean', 'sum'],
    'Rating'    : 'mean',
    'Invoice ID': 'count'
}).reset_index()

# Langkah ketiga kita mau meratakan nama kolom dari dataset yang sudah di agregasi ke dalam
# penamaan bahasa indonesia supaya lebih mudah dibacanya.
grouping_gender_customer.columns = ['Gender', 'Rata_Rata_Pengeluaran', 'Total_Pengeluaran',
                              'Rata_Rata_Rating', 'Jumlah_Transaksi']

# Langkah keempat kita mau menampilkan isi DataFrame dari grouping_gender_customer tanpa nomor index (index=False)
# dan menjadikan outputnya seperti tabel.
print("Tabel Analisis : \n")
print(grouping_gender_customer.to_string(index=False))

# Langkah kelima kita mau menampilkan hasil analisisnya.
print("\nHasil Analisis Minat Terhadap Produk (Health and Beauty) Pria VS Wanita: ")
if grouping_gender_customer.loc[grouping_gender_customer['Gender'] == 'Male', 'Rata_Rata_Pengeluaran'].values[0] > \
   grouping_gender_customer.loc[grouping_gender_customer['Gender'] == 'Female', 'Rata_Rata_Pengeluaran'].values[0]:
    print("Customer Pria Memiliki Minat Yang Lebih Besar Terhadap Produk (Health and Beauty) Daripada Customer Wanita")
else:
    print("Customer Wanita Memiliki Minat Yang Lebih Besar Terhadap Produk (Health and Beauty) Daripada Customer Pria")

Tabel Analisis : 

Gender  Rata_Rata_Pengeluaran  Total_Pengeluaran  Rata_Rata_Rating  Jumlah_Transaksi
Female             290.015414         18560.9865          7.146875                64
  Male             348.099460         30632.7525          6.898864                88

Hasil Analisis Minat Terhadap Produk (Health and Beauty) Pria VS Wanita: 
Customer Pria Memiliki Minat Yang Lebih Besar Terhadap Produk (Health and Beauty) Daripada Customer Wanita


# Hasil Kesimpulan:
- Proses Analisis Soal 3: Metode Grouping dan Aggregasi
- Pertanyaan Analisis: Berapa rata-rata pengeluaran dan rating terhadap produk (Health and Beauty) pada customer pria vs wanita? Aku ingin menganalisis produk (Health and Beauty) ini sebenarnya lebih diminati oleh customer pria atau wanita? Sudah sesuai dengan ketentuan soal dengan menggunakan kolom kategorikal "Genre" untuk komparasi customer pria vs wanita, kolom "Product Line" untuk pengujian terhadap produk (Health and Beauty), dan penggunaan kolom numerikal "Total" untuk pengujian terhadap data penjualan dan kolom numerikal "Rating" untuk mengukur kepuasan pelanggan.

1. Setelah melakukan proses analisis dengan metode grouping dan aggregasi terhadap pertanyaan analisis yang sudah dibuat sebelumnya, dapat ditarik kesimpulan bahwa customer pria memiliki minat yang lebih besar terhadap pembelian produk (Health and Beauty) dibandingkan dengan customer wanita. Hal tersebut menunjukkan insight yang unik dimana biasanya produk (Health and Beauty) identik dan familiar di kalangan customer wanita, tetapi dalam hal ini pembelian produknya justru di dominasi oleh customer pria. Saran dan masukan bagi pengelola supermarket seperti lebih meningkatkan stock terhadap produk (Health and Beauty) khusus pria dan juga menerapkan promo-promo menarik yang diharapkan bisa mendongkrak daya beli terhadap produk tersebut dan bisa meningkatkan tingkat kepuasan pelanggan terhadap perusahaan.

# Jawaban Soal 4: Pertanyaan Analisa Pada Metode Join / Merging

- Bagaimana performa dari tipe customer (Member vs Normal) di setiap cabang (branch) setelah digabungkan dengan kolom region pada dataset branch_info.csv dan kolom product line pada dataset product_info.csv? Apakah customer member di region/cabang tertentu mempunyai nilai rating atau total pengeluaran yang lebih tinggi jika dibandingkan dengan customer normal?

# Jawaban Soal 4: Coding Untuk Metode Join / Merging

In [10]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


In [11]:
data_cabang = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/branch_info.csv')
data_cabang.head()

,Branch,Region,Manager,Opened Year
0,A,South,Alice,2015
1,B,Central,Bob,2017
2,C,North,Charlie,2016


In [12]:
data_produk = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/product_info.csv')
data_produk.head()

,Product line,Category,Supplier,Warranty (months)
0,Health and beauty,Beauty,Supplier A,12
1,Electronic accessories,Electronics,Supplier B,24
2,Home and lifestyle,Home,Supplier C,18


In [13]:
# Langkah pertama kita mau melakukan merging data dengan menambahkan informasi dari kolom region pada DataFrame branch_info.
data_region = pd.merge(data_penjualan, data_cabang, on='Branch', how='left')

# Langkah kedua kita mau melanjutkan merging data sebelumnya dengan menambahkan informasi dari kolom Product line pada DataFrame product_info.
data_komplet = pd.merge(data_region, data_produk, on='Product line', how='left')

# Langkah ketiga kita mau melakukan grouping data berdasarkan Customer type, Branch, Region, dan Product line.
grouping_kolom = data_komplet.groupby(['Customer type', 'Branch', 'Region', 'Product line']).agg({
    'Total'      : 'sum',  # Total pengeluaran transaksi customer
    'Rating'     : 'mean', # Rata-rata rating customer
    'Invoice ID' : 'count' # Jumlah transaksi customer
}).reset_index()

# Langkah keempat kita mau meratakan nama kolom dari dataset yang sudah di agregasi ke dalam penamaan bahasa indonesia supaya lebih mudah dibacanya.
grouping_kolom.columns = ['Tipe_Customer', 'Cabang', 'Region', 'Product_Line',
                             'Total_Pengeluaran', 'Rata_Rata_Rating', 'Jumlah_Transaksi']

# Langkah kelima kita mau melakukan sorting untuk mengurutkan kolom Region secara ascending dan untuk kolom Total_Pengeluaran secara descending.
sorting_kolom = grouping_kolom.sort_values(by=['Region', 'Total_Pengeluaran'], ascending=[True, False])

# Langkah keenam kita mau menampilkan isi DataFrame dari sorting_kolom tanpa nomor index (index=False) dan menjadikan outputnya seperti tabel.
print("Tabel Analisis:\n")
print(sorting_kolom.to_string(index=False))

# Langkah ketujuh kita mau menampilkan hasil analisisnya.
print("\nHasil Analisis:")
data_member = sorting_kolom[sorting_kolom['Tipe_Customer'] == 'Member']
data_normal = sorting_kolom[sorting_kolom['Tipe_Customer'] == 'Normal']

if not data_member.empty and not data_normal.empty:
    if data_member['Total_Pengeluaran'].mean() > data_normal['Total_Pengeluaran'].mean():
        print("Customer Member secara keseluruhan memiliki total pengeluaran lebih tinggi dibandingkan Customer Normal di berbagai region/cabang.")
    else:
        print("Customer Normal secara keseluruhan memiliki total pengeluaran lebih tinggi dibandingkan Customer Member di berbagai region/cabang.")
    if data_member['Rata_Rata_Rating'].mean() > data_normal['Rata_Rata_Rating'].mean():
        print("Customer Member memiliki rata-rata rating lebih tinggi dibandingkan dengan Customer Normal di berbagai region/cabang.")
    else:
        print("Customer Normal memiliki rata-rata rating lebih tinggi dibandingkan dengan Customer Member di berbagai region/cabang.")

Tabel Analisis:

Tipe_Customer Cabang  Region           Product_Line  Total_Pengeluaran  Rata_Rata_Rating  Jumlah_Transaksi
       Member      B Central      Health and beauty         11327.8200          7.176923                26
       Member      B Central      Sports and travel         10420.0530          6.156667                30
       Normal      B Central     Home and lifestyle          9779.8575          6.344828                29
       Normal      B Central Electronic accessories          9626.5890          7.064286                28
       Normal      B Central      Sports and travel          9568.1460          6.840625                32
       Member      B Central     Food and beverages          9423.1200          7.024138                29
       Normal      B Central    Fashion accessories          9073.7850          7.003333                30
       Normal      B Central      Health and beauty          8652.8400          7.025926                27
       Member      B

# Hasil Kesimpulan:
- Proses Analisis Soal 4: Metode Join & Merging Data
- Pertanyaan Analisis: Bagaimana performa dari tipe customer (Member vs Normal) di setiap cabang (branch) setelah digabungkan dengan kolom region pada dataset branch_info.csv dan kolom product line pada dataset product_info.csv? Apakah customer member di region/cabang tertentu mempunyai nilai rating atau total pengeluaran yang lebih tinggi jika dibandingkan dengan customer normal?

1. Setelah melakukan proses analisis dengan metode join dan merging data terhadap pertanyaan analisis yang sudah dibuat sebelumnya, dapat ditarik kesimpulan bahwa proses merging dengan dataset branch_info.csv dan product_info.csv menghasilkan beberapa insight menarik, seperti customer member memberikan kontribusi terhadap total pengeluaran yang lebih tinggi dibandingkan dengan customer normal di berbagai region/cabang, hal tersebut menunjukkan bahwa program membership sukses mendorong jumlah transaksi produk yang lebih besar. Di sisi lain, customer normal memberikan kontribusi terhadap rata-rata pemberian rating yang lebih besar dibandingkan dengan customer member, hal tersebut menunjukkan bahwa customer normal mendapatkan pengalaman pelayanan yang lebih baik dibandingkan customer member. Saran dan masukan bagi pengelola supermarket untuk meningkatkan mutu layanan yang lebih baik dan diharapkan bisa merata supaya bisa dirasakan oleh semua tipe customer baik yang member maupun yang tidak member (normal), sehingga kedepannya dapat meningkatkan nilai pelayanan yang lebih baik dan dibarengin dengan tingkat pendapatan yang lebih tinggi bagi perusahaan.

# Jawaban Soal 5: Pertanyaan Analisa Pada Metode Pivot

- Bagaimana distribusi data yang paling banyak terjual selama (1 bulan terakhir) pada Product line? Apakah terdapat perbedaan dalam produk terlaris di setiap cabang tertentu? Aku mengerjakannya sesuai ketentuan konteks soal dengan memfilter data hanya untuk 1 bulan terakhir dan mengekstrak data bulan dari kolom Date, selanjutnya membuat pivot table dengan Product line sebagai baris, Branch sebagai kolom, dan Quantity sebagai nilai fungsi agregasi sum.

# Jawaban Soal 5: Coding Untuk Metode Pivot

In [14]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


In [15]:
# Langkah pertama kita mau melakukan konversi terhadap kolom Date ke dalam format datetime untuk filter dan analisis waktu.
data_penjualan['Date'] = pd.to_datetime(data_penjualan['Date'])

# Langkah kedua kita mau melakukan pencarian tanggal maksimum untuk menentukan data dari 1 bulan terakhir.
tanggal_maksimum = data_penjualan['Date'].max()
tanggal_minimum_bulan_terakhir = tanggal_maksimum - pd.Timedelta(days=30)

# Langkah ketiga kita mau melakukan filtering data hanya untuk 1 bulan terakhir berdasarkan tanggal.
data_bulan_terakhir = data_penjualan[data_penjualan['Date'] >= tanggal_minimum_bulan_terakhir]

# Langkah keempat kita mau membuat tabel pivot untuk Sum Quantity (total kuantitas) dengan Product line sebagai baris dan Branch sebagai kolom.
tabel_pivot = pd.pivot_table(
    data_bulan_terakhir,
    values='Quantity',       # Kolom numerikal (Quantity untuk hitung total terjual)
    index='Product line',    # Kolom kategorikal (Product line sebagai baris)
    columns='Branch',        # Kolom kategorikal (Branch sebagai kolom)
    aggfunc='sum'            # Fungsi agregasi (sum untuk total quantity)
)

# Langkah kelima kita mau menampilkan hasil pivot tabelnya.
print("Pivot Table: Total Quantity per Product Line yang Terjual di Cabang Tertentu Selama (1 Bulan Terakhir)\n")
print(tabel_pivot)

# Langkah keenam kita mau menampilkan hasil analisisnya dengan mencari Product line paling laris di cabang tertentu.
print("\nHasil Analisis:")
for cabang in tabel_pivot.columns:
    produk_terlaris = tabel_pivot[cabang].idxmax()
    kuantitas_maksimum = tabel_pivot[cabang].max()
    print(f"Cabang {cabang}: Product line yang paling banyak terjual adalah '{produk_terlaris}' dengan total {kuantitas_maksimum} unit.")

Pivot Table: Total Quantity per Product Line yang Terjual di Cabang Tertentu Selama (1 Bulan Terakhir)

Branch                    A    B    C
Product line                         
Electronic accessories  116   80  138
Fashion accessories      70   86  127
Food and beverages      102   61  121
Health and beauty       135  127   75
Home and lifestyle      128  139   97
Sports and travel       139  114   66

Hasil Analisis:
Cabang A: Product line yang paling banyak terjual adalah 'Sports and travel' dengan total 139 unit.
Cabang B: Product line yang paling banyak terjual adalah 'Home and lifestyle' dengan total 139 unit.
Cabang C: Product line yang paling banyak terjual adalah 'Electronic accessories' dengan total 138 unit.


# Hasil Kesimpulan:
- Proses Analisis Soal 5: Metode Pivot Data
- Pertanyaan Analisis: Bagaimana distribusi data yang paling banyak terjual selama (1 bulan terakhir) pada Product line? Apakah terdapat perbedaan dalam produk terlaris di setiap cabang tertentu? Aku mengerjakannya sesuai ketentuan konteks soal dengan memfilter data hanya untuk 1 bulan terakhir dan mengekstrak data bulan dari kolom Date, selanjutnya membuat pivot table dengan Product line sebagai baris, Branch sebagai kolom, dan Quantity sebagai nilai fungsi agregasi sum.

1. Setelah melakukan proses analisis dengan metode pivot data terhadap pertanyaan analisis yang sudah dibuat sebelumnya, dapat ditarik kesimpulan bahwa cabang supermarket A produk yang paling banyak terjual adalah (Sports and Travel) dengan 139 unit terjual, cabang supermarket B dengan produk (Home and Lifestyle) dengan 139 unit terjual, dan cabang supermarket C dengan produk (Electronic Accessories) dengan 138 unit terjual. Dari hasil analisis data tersebut terdapat saran dan masukan bagi pengelola supermarket seperti menambahkan stok pada produk terlaris di masing-masing cabang yang penjualan terhadap produk tersebut rendah, sehingga bisa berkontribusi dalam total penjualan dan menambah jumlah pendapatan perusahaan.

# Jawaban Soal 6: Pertanyaan Analisa Pada Metode Crosstab

- Bagaimana frekuensi penggunaan metode pembayaran di setiap cabang? Apakah ada cabang yang lebih banyak menggunakan Credit Card dibandingkan metode lain, sehingga bisa diberikan rekomendasi penggunaan E-Wallet untuk opsi pengganti dari Credit Card? Aku sudah menjawab sesuai dengan ketentuan soal dengan melakukan crosstab antara kolom 'Payment' dengan kolom 'Branch'.

# Jawaban Soal 6: Coding Untuk Metode Crosstab

In [16]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


In [17]:
# Langkah pertama kita mau membuat tabel crosstab untuk mengukur frekuensi antara metode pembayaran (Payment) dengan cabang (Branch).
tabel_crosstab = pd.crosstab(
    index=data_penjualan['Payment'], # Parameter disamping ini menentukan bahwa metode pembayaran (Credit Card, E-Wallet, dan Cash) menjadi baris pada tabel.
    columns=data_penjualan['Branch'], # Parameter disamping ini menentukan bahwa cabang supermarket (A, B, C) menjadi kolom pada tabel.
    margins=True, # Parameter di samping ini memerintahkan Pandas menambahkan baris dan kolom di bagian bawah pada sisi kiri dan kanan tabel.
    margins_name='Total' # Parameter di samping ini untuk memberikan nama 'Total' pada baris dan kolom yang sudah ditambahkan sebelumnya.
)

# Langkah kedua kita mau menampilkan hasil tabel crosstab yang sebelumnya sudah dibuat.
print("Table Crosstab: Frekuensi Penggunaan Metode Pembayaran per Cabang\n")
print(tabel_crosstab)

# Langkah ketiga kita mau menghitung persentase dari metode pembayaran (Credit card) pada tiap cabang dengan cara mengambil baris 'Credit card' saja tanpa total.
proporsi_credit_card = tabel_crosstab.loc['Credit card'] / tabel_crosstab.loc['Total'] * 100

# Langkah keempat kita mau menghitung persentase dari metode pembayaran (E-Wallet).
proporsi_ewallet = tabel_crosstab.loc['Ewallet'] / tabel_crosstab.loc['Total'] * 100

# Langkah kelima kita mau menampilkan persentase dari (Credit card) dan (E-Wallet) pada tiap cabang.
print("\nTable Persentase Perbandingan (Credit Card) VS (E-Wallet):")
print("Persentase penggunaan Credit Card di setiap cabang:")
print(proporsi_credit_card.round(2).astype(str) + '%')

print("\nPersentase penggunaan E-Wallet di setiap cabang:")
print(proporsi_ewallet.round(2).astype(str) + '%')

# Langkah keenam kita mau menampilkan hasil analisis berdasarkan frekuensi penggunaan metode (Credit card).
cabang_tertinggi_credit = proporsi_credit_card.idxmax()
persentase_tertinggi_credit = proporsi_credit_card.max()

print(f"\nTable Hasil Analisis:")
print(f"Cabang dengan penggunaan Credit card paling tinggi adalah cabang {cabang_tertinggi_credit} dengan persentase sebesar {persentase_tertinggi_credit:.2f}%.")

Table Crosstab: Frekuensi Penggunaan Metode Pembayaran per Cabang

Branch         A    B    C  Total
Payment                          
Cash         110  110  124    344
Credit card  104  109   98    311
Ewallet      126  113  106    345
Total        340  332  328   1000

Table Persentase Perbandingan (Credit Card) VS (E-Wallet):
Persentase penggunaan Credit Card di setiap cabang:
Branch
A        30.59%
B        32.83%
C        29.88%
Total     31.1%
dtype: object

Persentase penggunaan E-Wallet di setiap cabang:
Branch
A        37.06%
B        34.04%
C        32.32%
Total     34.5%
dtype: object

Table Hasil Analisis:
Cabang dengan penggunaan Credit card paling tinggi adalah cabang B dengan persentase sebesar 32.83%.


# Hasil Kesimpulan:
- Proses Analisis Soal 6: Metode Crosstab Analysis
- Pertanyaan Analisis: Bagaimana frekuensi penggunaan metode pembayaran di setiap cabang? Apakah ada cabang yang lebih banyak menggunakan Credit Card dibandingkan metode lain, sehingga bisa diberikan rekomendasi penggunaan E-Wallet untuk opsi pengganti dari Credit Card? Aku sudah menjawab sesuai dengan ketentuan soal dengan melakukan crosstab antara kolom 'Payment' dengan kolom 'Branch'.

1. Setelah melakukan proses analisis dengan metode crosstab analysis terhadap pertanyaan analisis yang sudah dibuat sebelumnya, dapat ditarik kesimpulan bahwa terdapat variasi penggunaan metode pembayaran di setiap cabang supermarket. Cabang dengan penggunaan metode pembayaran (Credit Card) terbanyak adalah cabang B dengan persentase sebesar 32.83%, sedangkan cabang dengan penggunaan metode pembayaran (E-Wallet) terbanyak adalah cabang A dengan persentase sebesar 37.06%. Hal tersebut bisa diberikan saran dan masukan bagi pengelola supermarket seperti melakukan promosi terkait penggunaan (E-Wallet) di cabang-cabang yang masih di dominasi oleh pengguna (Credit Card), strategi promosinya bisa dengan memberikan diskon spesial jika bertransaksi dengan (E-Wallet). Diharapkan dengan promosi tersebut dapat memberikan pengalaman bertransaksi yang mudah bagi pelanggan dengan ditambahnya metode pembayaran (E-Wallet) di cabang-cabang yang di dominasi oleh pengguna (Credit Card) selain itu bisa berpotensi menurunkan beban biaya administrasi transaksi bagi pengguna, karena biaya admin untuk (E-Wallet) jauh lebih rendah jika dibandingkan dengan (Credit Card).

# Jawaban Soal 7: Pertanyaan Analisa Pada Metode Appending

- (Tulis Jawabannya Disini)

# Jawaban Soal 7: Coding Untuk Metode Appending

# Jawaban Soal 8: Pertanyaan Analisa Pada Metode Mengeksplor terkait dengan operasi pada Datetime (week, day, dan sebagainya)

- Bagaimana distribusi total penjualan berdasarkan periode waktu harian dan (Product line)? Apakah terdapat periode waktu tersibuk (peak hour) berdasarkan total penjualan untuk mengukur produk apa saja yang paling laris pada periode tersebut? Aku mengerjakan soal ini sesuai dengan ketentuan yaitu menggunakan operasi datetime untuk mengkonversi kolom Time menjadi jam harian, lalu mengklompokkannya menjadi periode waktu (pagi, siang, sore).

# Jawaban Soal 8: Coding Untuk Metode Mengeksplor terkait dengan operasi pada Datetime (week, day, dan sebagainya)

In [18]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


In [19]:
# Langkah pertama kita mau mengkonversi kolom Time ke dalam datetime.
data_penjualan['Time'] = pd.to_datetime(data_penjualan['Time'], format='%H:%M')

# Langkah kedua kita mau mengekstrak kolom Time menjadi jam harian.
data_penjualan['Jam'] = data_penjualan['Time'].dt.hour

# Langkah ketiga kita mau mengklompokkan menjadi periode waktu (Pagi, Siang, Sore).
def kelompok_periode(jam):
    if 9 <= jam < 13:
        return 'Pagi (09:00 - 12:59)'
    elif 13 <= jam < 17:
        return 'Siang (13:00 - 16:59)'
    elif 17 <= jam < 21:
        return 'Sore (17:00 - 20:59)'
    else:
        return 'Malam (21:00 - 08:59)'

data_penjualan['Periode_Waktu'] = data_penjualan['Jam'].apply(kelompok_periode)

# Langkah keempat kita mau menghitung total penjualan berdasarkan periode waktu harian dan Product line.
total_periode_produk = data_penjualan.groupby(['Periode_Waktu', 'Product line'])['Total'].sum().reset_index()

# Langkah kelima kita mau menghitung total penjualan per periode waktu harian untuk menemukan waktu tersibuk.
total_periode_tersibuk = total_periode_produk.groupby('Periode_Waktu')['Total'].sum().reset_index()
total_periode_tersibuk = total_periode_tersibuk.sort_values('Total', ascending=False)

# Langkah keenam kita mau mengurutkan periode waktu harian untuk nantinya ditampilkan datanya.
urutan_periode = ['Pagi (09:00 - 12:59)', 'Siang (13:00 - 16:59)',
                  'Sore (17:00 - 20:59)', 'Malam (21:00 - 08:59)']
total_periode_produk['Periode_Waktu'] = pd.Categorical(total_periode_produk['Periode_Waktu'],
                                                 categories=urutan_periode, ordered=True)
analisis_sorted = total_periode_produk.sort_values(['Periode_Waktu', 'Total'], ascending=[True, False])

# Langkah ketujuh kita mau menampilkan detail tabel terkait penjualan berdasarkan periode waktu harian dan total penjualan beserta list per Product line.
print("Tabel Analisis Penjualan per Periode Waktu dan Lini Produk\n")
print(analisis_sorted.to_string(index=False))

# Langkah kedelapan kita mau menampilkan hasi analisisnya dengan cara mencari Product line paling laris di periode waktu tersibuk.
print("\nHasil Analisis:")
if not total_periode_tersibuk.empty:
    periode_sibuk = total_periode_tersibuk.iloc[0]['Periode_Waktu']
    total_penjualan_sibuk = total_periode_tersibuk.iloc[0]['Total']

    sorting_periode_sibuk = analisis_sorted[analisis_sorted['Periode_Waktu'] == periode_sibuk]
    produk_dominan = sorting_periode_sibuk.iloc[0]['Product line']
    total_produk_laris = sorting_periode_sibuk.iloc[0]['Total']

    print(f"Periode waktu tersibuk adalah '{periode_sibuk}' dengan total penjualan: {total_penjualan_sibuk:.2f} unit.")
    print(f"Product line paling laris pada periode waktu tersebut adalah '{produk_dominan}' dengan total penjualan: {total_produk_laris:.2f} unit.")

Tabel Analisis Penjualan per Periode Waktu dan Lini Produk

        Periode_Waktu           Product line      Total
 Pagi (09:00 - 12:59)     Home and lifestyle 16029.7620
 Pagi (09:00 - 12:59)     Food and beverages 15571.4475
 Pagi (09:00 - 12:59)    Fashion accessories 15422.1270
 Pagi (09:00 - 12:59)      Sports and travel 14551.3830
 Pagi (09:00 - 12:59) Electronic accessories 13610.9925
 Pagi (09:00 - 12:59)      Health and beauty 12678.9810
Siang (13:00 - 16:59)      Sports and travel 22810.4520
Siang (13:00 - 16:59)    Fashion accessories 21856.6215
Siang (13:00 - 16:59)     Food and beverages 19991.8950
Siang (13:00 - 16:59)      Health and beauty 19568.5770
Siang (13:00 - 16:59)     Home and lifestyle 18936.3405
Siang (13:00 - 16:59) Electronic accessories 18793.5720
 Sore (17:00 - 20:59) Electronic accessories 21932.9670
 Sore (17:00 - 20:59)     Food and beverages 20581.5015
 Sore (17:00 - 20:59)     Home and lifestyle 18895.8105
 Sore (17:00 - 20:59)      Sports and travel

# Hasil Kesimpulan:
- Proses Analisis Soal 8: Metode Datetime Operations  
- Pertanyaan Analisis: Bagaimana distribusi total penjualan berdasarkan periode waktu harian dan (Product line)? Apakah terdapat periode waktu tersibuk (peak hour) berdasarkan total penjualan untuk mengukur produk apa saja yang paling laris pada periode tersebut? Aku mengerjakan soal ini sesuai dengan ketentuan yaitu menggunakan operasi datetime untuk mengkonversi kolom Time menjadi jam harian, lalu mengklompokkannya menjadi periode waktu (pagi, siang, sore).

1. Setelah melakukan proses analisis dengan metode datetime operations  terhadap pertanyaan analisis yang sudah dibuat sebelumnya, dapat ditarik kesimpulan bahwa periode waktu tersibuk pada supermarket tersebut terdapat di waktu siang pada jam (13:00 - 16:59) dengan total penjualan keseluruhannya adalah 121957.46 unit. Untuk bagian Product line paling laris pada periode waktu tersebut adalah produk (Sports and travel) dengan total penjualan keseluruhannya adalah 22810.45 unit. Dari hasil analisis keseluruhan tersebut terdapat saran dan masukan bagi pengelola supermarket, seperti memberikan paket diskon menarik khusus untuk pembelian produk (Sports and travel) di waktu siang pada jam (13:00 - 16:59), diharapkan dari strategi promosi seperti itu bisa meningkatkan nilai kepuasan pelanggan dan juga dapat meningkatkan revenue pada perusahaan.

# 9. Standar Scaller for Exploratory Data Analysis ?
`clue` : lakukan tanpa splitting dataset ke train dan test
- lakukan pada kolom Total, Unit Price, dan tambahkan 2 kolom lainnya
- dan tulis insight dari yang anda dapat dari grafik tersebut

# Jawaban Soal 9: Pertanyaan Analisa Pada Metode Standar Scaller for Exploratory Data Analysis

- Aku ingin memilih 4 kolom numerik sesuai dengan ketentuan konteks soal, yaitu kolom Total, cogs (Cost of Goods Sold atau Harga Pokok Penjualan (HPP)), gross income (Pendapatan Kotor), dan Rating. Selanjutnya menerapkan Standar Scaller untuk proses Exploratory Data Analysis (EDA). Lalu, bagaimana nilai-nilai dari variabel keuangan (total penjualan, harga pokok penjualan, dan pendapatan kotor) dapat dibandingkan dengan tingkat Rating? Apakah dari proses tersebut dapat diperoleh sebuah informasi terkait hubungan dari performa penjualan dan pengalaman pengguna setelah dilakukan normalisasi?

# Jawaban Soal 9: Coding Untuk Metode Standar Scaller for Exploratory Data Analysis

In [20]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


In [23]:
from sklearn.preprocessing import StandardScaler

# Langkah pertama kita mau melakukan pemilihan untuk 4 kolom numerik yang bisa untuk di scaling dan juga sesuai dengan konteks soal.
kolom_numerik = ['Total', 'cogs', 'gross income', 'Rating']

# Langkah kedua kita mau melakukan pembuatan DataFrame baru yang hanya diisi oleh 4 kolom numerik yang sudah dijelaskan sebelumnya.
data_numerik = data_penjualan[kolom_numerik].copy()

# Langkah ketiga kita mau melakukan inisialisasi objek StandardScaler dari scikit-learn.
scaler = StandardScaler()

# Langkah keempat kita mau menerapkan scaling pada 4 kolom numerik.
data_scaled = scaler.fit_transform(data_numerik)

# Langkah kelima kita mau mengubah hasil scaling menjadi DataFrame supaya mudah saat dibaca.
data_scaled_df = pd.DataFrame(data_scaled, columns=kolom_numerik)

# Langkah keenam kita mau menampilkan hasil tabel analisis untuk melihat perubahan data setelah proses scaling.
print("Tabel Analisis Hasil Scaling Dengan Standard Scaler (10 Baris Pertama):")
print(data_scaled_df.head(10).round(4))

# Langkah ketujuh kita mau menampilkan hasil analisis untuk deskriptif statistik setelah dilakukan proses scaling.
print("\nStatistik Deskriptif Setelah Scaling:")
print(data_scaled_df.describe().round(4))

Tabel Analisis Hasil Scaling Dengan Standard Scaler (10 Baris Pertama):
    Total    cogs  gross income  Rating
0  0.9196  0.9196        0.9196  1.2384
1 -0.9877 -0.9877       -0.9877  1.5295
2  0.0714  0.0714        0.0714  0.2488
3  0.6758  0.6758        0.6758  0.8309
4  1.2671  1.2671        1.2671 -0.9738
5  1.2396  1.2396        1.2396 -1.6724
6  0.4505  0.4505        0.4505 -0.6827
7  1.8286  1.8286        1.8286  0.5981
8 -1.0043 -1.0043       -1.0043  0.1323
9 -0.6112 -0.6112       -0.6112 -0.6245

Statistik Deskriptif Setelah Scaling:
           Total       cogs  gross income     Rating
count  1000.0000  1000.0000     1000.0000  1000.0000
mean      0.0000     0.0000        0.0000     0.0000
std       1.0005     1.0005        1.0005     1.0005
min      -1.2707    -1.2707       -1.2707    -1.7306
25%      -0.8079    -0.8079       -0.8079    -0.8574
50%      -0.2812    -0.2812       -0.2812     0.0159
75%       0.6038     0.6038        0.6038     0.8891
max       2.9284     2.92

# Hasil Kesimpulan:
- Proses Analisis Soal 9: Metode Standar Scaller for Exploratory Data Analysis  
- Pertanyaan Analisis: Aku ingin memilih 4 kolom numerik sesuai dengan ketentuan konteks soal, yaitu kolom Total, cogs (Cost of Goods Sold atau Harga Pokok Penjualan (HPP)), gross income (Pendapatan Kotor), dan Rating. Selanjutnya menerapkan Standar Scaller untuk proses Exploratory Data Analysis (EDA). Lalu, bagaimana nilai-nilai dari variabel keuangan (total penjualan, harga pokok penjualan, dan pendapatan kotor) dapat dibandingkan dengan tingkat Rating? Apakah dari proses tersebut dapat diperoleh sebuah informasi terkait hubungan dari performa penjualan dan pengalaman pengguna setelah dilakukan normalisasi?

1. Setelah melakukan proses analisis dengan metode standar scaller for exploratory data analysis terhadap pertanyaan analisis yang sudah dibuat sebelumnya, dapat ditarik kesimpulan bahwa ketiga variabel keuangan yang terdiri dari Total, cogs, dan gross income didapatkan nilai distribusi yang identik setelah dilakukan proses scaling dengan mean sama dengan 0 dan standar deviasi sama dengan 1, dibarengi dengan range nya di sekitar -1.27 hingga +2.93, sementara kolom Rating didapatkan variasi yang lebih sempit di sekitar -1.73 hingga +1.76, sehingga dapat disimpulkan bahwa variasi pada variabel keuangan nilainya yang lebih besar dibandingkan tingkat kepuasan pelanggan, Saran dan masukan bagi pengelola supermarket agar menjadikan proses analisis ini bisa dijadikan acuan untuk melakukan proses evaluasi untuk bisa lebih memprioritaskan peningkatan pengalaman pelanggan dan untuk bisa menyeimbangkan antara prioritas menigkatkan pendapatan perusahaan dengan menigkatkan pengalaman dan kepuasan pengguna.

# 10. MinMax Scaller for Exploratory Data Analysis ?
`clue` : lakukan tanpa splitting dataset ke train dan test
- lakukan pada kolom Total, Unit Price, dan tambahkan 2 kolom lainnya
- dan tulis insight dari yang anda dapat dari grafik tersebut

# Jawaban Soal 10: Pertanyaan Analisa Pada Metode MinMax Scaller for Exploratory Data Analysis

# Jawaban Soal 10: Coding Untuk Metode MinMax Scaller for Exploratory Data Analysis

In [22]:
data_penjualan = pd.read_csv('/content/drive/MyDrive/Assignment Day 13 Data Manipulation_Raihan Azhar Rafi/supermarket_sales.csv')
data_penjualan.head(10)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
5,699-14-3026,C,Naypyitaw,Normal,Male,Electronic accessories,85.39,7,29.8865,627.6165,3/25/2019,18:30,Ewallet,597.73,4.761905,29.8865,4.1
6,355-53-5943,A,Yangon,Member,Female,Electronic accessories,68.84,6,20.6520,433.6920,2/25/2019,14:36,Ewallet,413.04,4.761905,20.6520,5.8
7,315-22-5665,C,Naypyitaw,Normal,Female,Home and lifestyle,73.56,10,36.7800,772.3800,2/24/2019,11:38,Ewallet,735.60,4.761905,36.7800,8.0
8,665-32-9167,A,Yangon,Member,Female,Health and beauty,36.26,2,3.6260,76.1460,1/10/2019,17:15,Credit card,72.52,4.761905,3.6260,7.2
9,692-92-5582,B,Mandalay,Member,Female,Food and beverages,54.84,3,8.2260,172.7460,2/20/2019,13:27,Credit card,164.52,4.761905,8.2260,5.9


# Hasil Kesimpulan:
- Proses Analisis Soal 10: Metode MinMax Scaller for Exploratory Data Analysis  
- Pertanyaan Analisis: